include/aspect/particle/world.h

```diff
@@ -376,14 +369,19 @@ namespace aspect
                                  const typename parallel::distributed::Triangulation<dim>::active_cell_iterator &cell) const;
 
         /**
-         * Finds the cells containing each particle for all particles. If
-         * particles moved out of this subdomain they will be sent
-         * to their new process and inserted there. After this function call
-         * every particle is either on its current process and in its current
-         * cell, or deleted (if it could not find its new process or cell).
+         * Finds the cells containing each particle for all particles in
+         * @p particles_to_sort. If particles moved out of the local subdomain
+         * they will be sent to their new process and inserted there.
+         * After this function call every particle is either on its current
+         * process and in its current cell, or deleted (if it could not find
+         * its new process or cell).
+         *
+         * @param [in,out] particles_to_sort Multimap containing all particles
+         * that will be sorted in this function. After the function call this
+         * container is empty.
          */
         void
-        sort_particles_in_subdomains_and_cells();
+        sort_particles_in_subdomains_and_cells(std::list<std::pair<types::LevelInd, Particle<dim> > > &particles_to_sort);
```

Comment:
I think it's probably going to be in the noise performancewise since the Newton iteration is so expensive. In general, a list is a pretty expensive data structure since it needs to store pointers to the previous and next elements. It also fragments the memory. In practice, I imagine that `std::list` is implemented via buffers of multiple elements and memory pools, so it probably doesn't make all that much of a difference. If it's easy to try out, go ahead. Otherwise, don't worry about it.

(As for appending elements to a vector that already has enough memory allocated: A `std::vector` stores a pointer to the beginning of its memory range, an integer for the number of allocated elements, and a number of used elements. Appending an element is extremely cheap if there are unused elements at the end of the vector. It only becomes expensive if the allocated memory is not enough, since then you need to copy the entire existing vector. To make this as infrequent as possible, whenever you append an element and memory needs to be allocated, typical `std::vector` implementations allocate twice as much memory as before. So you quickly grow to a size that is sufficient for everything you want to do, without having to do too many re-allocations/copy operations. But you can help if you call `std::vector::reserve` at the beginning to already allocate what you think is a reasonable guess for the number of memory slots you might need; underestimating does not harm, just causes re-allocation and copying later.)